plugins/module_utils/cloud_py.py

```diff
@@ -0,0 +1,185 @@
+# Copyright (c) 2021 Ansible Project
+#
+# This code is part of Ansible, but is an independent component.
+# This particular file snippet, and this file snippet only, is BSD licensed.
+# Modules you write using this snippet, which is embedded dynamically by Ansible
+# still belong to the author of the module, and may assign their own license
+# to the complete work.
+#
+# Redistribution and use in source and binary forms, with or without modification,
+# are permitted provided that the following conditions are met:
+#
+#    * Redistributions of source code must retain the above copyright
+#      notice, this list of conditions and the following disclaimer.
+#    * Redistributions in binary form must reproduce the above copyright notice,
+#      this list of conditions and the following disclaimer in the documentation
+#      and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import time
+import functools
+import random
+
+
+def _backoff_value_generator(delay, backoff, max_delay=None, jitter=False):
+    """Generates sleep value based on the exponential or jitter back-off algorithm.
+    Args:
+        delay (int or float): initial delay.
+        backoff (int or float): backoff multiplier e.g. value of 2 will  double the delay each retry.
+        max_delay (int or None): maximum amount of time to wait between retries.
+        jitter (bool): if set to true, add jitter to the generate value.
+    Yields:
+        float: successive sleep value.
+    """
+    current_delay = delay
+    while True:
+        sleep_time = current_delay
+        return_value = sleep_time if max_delay is None else min(sleep_time, max_delay)
+        if jitter:
+            return_value = random.uniform(0.0, return_value)
+        yield return_value
```

Comment:
Caveat: It's possible I've misunderstood the exact scoping of the generator created here when used inside a decorator.

Given we yield here, rather than being a class that implements `__iter__` and `__next__` I believe this is going to mean that the backoff doesn't reset when we make a new call to the same function.  I suspect this isn't the behaviour people would anticipate.

Reply:
For example, in the AWS collections we sometimes retry while waiting on something to finish updating before making a second change to the resource (easier than going through the hassle of a full waiter).

If multiple parameters are updated by the same API call, we could end up waiting 10 seconds the first time, then straight away waiting 20 seconds for the second parameter, getting exponentially worse as we update each parameter.

Reply:
Each time we are running the ``_retry_func`` it will go through a new for loop which will start generator from the init value

Reply:
Could you do me a favour and add a unit test just to make sure that re-entering a wrapped function restarts the generator.

Reply:
@tremble you were definitely right, using a generator does not reset the value